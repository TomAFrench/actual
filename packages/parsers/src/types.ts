export type Transaction = {
  amount: number | null;
  date: string;
  payee_name: string;
  imported_payee: string;
  notes: string | null;
};

export type ParserOptions = CSVParserOptions | undefined;

export type CSVParserOptions = {
  delimiter: string;
  singleAmountField: boolean;
  headings: Record<keyof Omit<Transaction, 'amount'>, string> & { amount?: string; deposit?: string; payment?: string };
};

export type ParsingError = {
  message: string;
  internal: string;
};
