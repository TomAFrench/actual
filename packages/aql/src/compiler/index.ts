export { compileQuery, generateSQL, generateSQLWithState } from "./compiler";
