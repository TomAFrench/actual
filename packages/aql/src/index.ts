export { compileQuery, generateSQL, generateSQLWithState } from "./compiler";
export { q, Query, getPrimaryOrderBy } from "./query"
