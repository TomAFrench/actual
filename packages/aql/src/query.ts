
type TableOptions = {
  splits: "all" | "grouped" | "inline" | "none"
}

export type QueryState = {
    filterExpressions: any[];
    selectExpressions: any[];
    groupExpressions: any[];
    orderExpressions: any[];
    limit: number | null;
    offset: number | null;
    calculation: boolean;
    rawMode: boolean;
    withDead: boolean;
    validateRefs: boolean;
    table?: string;
    tableOptions?: TableOptions;
}

export class Query {
  state: QueryState = {
    filterExpressions: [],
    selectExpressions: [],
    groupExpressions: [],
    orderExpressions: [],
    calculation: false,
    rawMode: false,
    withDead: false,
    validateRefs: true,
    limit: null,
    offset: null,
  };

  constructor(state: Partial<QueryState>) {
    this.state = {
      ...this.state,
      ...state
    };
  }

  filter(expr): Query {
    return new Query({
      ...this.state,
      filterExpressions: [...this.state.filterExpressions, expr]
    });
  }

  unfilter(exprs): Query {
    let exprSet = new Set(exprs);
    return new Query({
      ...this.state,
      filterExpressions: this.state.filterExpressions.filter(
        expr => !exprSet.has(Object.keys(expr)[0])
      )
    });
  }

  select(exprs: any = []): Query {
    if (!Array.isArray(exprs)) {
      exprs = [exprs];
    }

    let query = new Query({ ...this.state, selectExpressions: exprs });
    query.state.calculation = false;
    return query;
  }

  calculate(expr): Query {
    let query = this.select({ result: expr });
    query.state.calculation = true;
    return query;
  }

  groupBy(exprs): Query {
    if (!Array.isArray(exprs)) {
      exprs = [exprs];
    }

    return new Query({
      ...this.state,
      groupExpressions: [...this.state.groupExpressions, ...exprs]
    });
  }

  orderBy(exprs): Query {
    if (!Array.isArray(exprs)) {
      exprs = [exprs];
    }

    return new Query({
      ...this.state,
      orderExpressions: [...this.state.orderExpressions, ...exprs]
    });
  }

  limit(limit: number): Query {
    return new Query({ ...this.state, limit });
  }

  offset(offset: number): Query {
    return new Query({ ...this.state, offset });
  }

  raw(): Query {
    return new Query({ ...this.state, rawMode: true });
  }

  withDead(): Query {
    return new Query({ ...this.state, withDead: true });
  }

  withoutValidatedRefs(): Query {
    return new Query({ ...this.state, validateRefs: false });
  }

  options(tableOptions: TableOptions): Query {
    return new Query({ ...this.state, tableOptions });
  }

  serialize(): QueryState {
    return this.state;
  }
}

export function getPrimaryOrderBy(query: Query, defaultOrderBy: {field: string; order?: "asc" | "desc"}): {field: string; order?: "asc" | "desc"} {
  let orderExprs = query.serialize().orderExpressions;
  if (orderExprs.length === 0) {
    if (defaultOrderBy) {
      return { order: 'asc', ...defaultOrderBy };
    }
    return null;
  }

  let firstOrder = orderExprs[0];
  if (typeof firstOrder === 'string') {
    return { field: firstOrder, order: 'asc' };
  }
  // Handle this form: { field: 'desc' }
  let [field] = Object.keys(firstOrder);
  return { field, order: firstOrder[field] };
}

export const q = (table: string) =>  new Query({ table });
