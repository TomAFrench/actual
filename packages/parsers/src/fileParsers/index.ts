export { parseCSV } from './csv';
export { parseQIF } from './qif';
export { parseOFX } from './ofx';
