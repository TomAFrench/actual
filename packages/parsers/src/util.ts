// This is used when the input format could be anything (from
// financial files and we don't want to parse based on the user's
// number format, because the user could be importing from many
// currencies. We extract out the numbers and just ignore separators.
export function looselyParseAmount(amount: string): number {
  function safeNumber(v) {
    return isNaN(v) ? null : v;
  }

  const m = amount.match(/[.,][^.,]*$/);
  if (!m || m.index === 0) {
    return safeNumber(parseFloat(amount));
  }

  const left = amount.slice(0, m.index);
  const right = amount.slice(m.index + 1);

  return safeNumber(parseFloat(left.replace(/[^0-9-]/g, '') + '.' + right));
}
