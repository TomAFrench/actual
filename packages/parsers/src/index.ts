import fs from 'fs/promises';
import { parseCSV, parseOFX, parseQIF } from './fileParsers';
import { CSVParserOptions, ParserOptions, ParsingError, Transaction } from './types';

const VALID_EXTENSIONS = ['.qif', '.csv', '.ofx', '.qfx'] as const;
type ValidExtension = typeof VALID_EXTENSIONS[number];

export const extractFileExtension = (filepath: string): string | null => {
  if (typeof filepath !== 'string') return null;
  const match = filepath.match(/\.[^.]*$/);
  if (match === null) return null;

  return match[0].toLowerCase();
};

export async function parseFile(
  filepath: string,
  options?: ParserOptions
): Promise<{ errors: ParsingError[]; transactions: Transaction[] }> {
  const extension = extractFileExtension(filepath);

  if (VALID_EXTENSIONS.includes(extension as ValidExtension)) {
    const fileContents = await fs.readFile(filepath);

    switch (extension) {
      case '.qif':
        return parseQIF(fileContents);
      case '.csv':
        return parseCSV(fileContents, options as CSVParserOptions);
      case '.ofx':
      case '.qfx':
        return parseOFX(fileContents);
      default:
    }
  }

  return {
    errors: [
      {
        message: 'Invalid file type',
        internal: '',
      },
    ],
    transactions: [],
  };
}
