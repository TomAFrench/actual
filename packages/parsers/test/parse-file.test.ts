import { expect } from '@jest/globals';
import { extractFileExtension, parseFile } from '../src';

describe('parseFile', () => {
  it('csv import works', async () => {
    const filePath = __dirname + '/data/data.csv';
    const { errors, transactions } = await parseFile(filePath, {
      delimiter: ',',
      singleAmountField: false,
      headings: {
        deposit: 'Credit Amount',
        payment: 'Debit Amount',
        date: 'Transaction Date',
        payee_name: 'Transaction Description',
        imported_payee: 'Transaction Description',
        notes: '',
      },
    });
    expect(errors.length).toBe(0);
    expect(transactions).toMatchSnapshot();
  });

  it('qif import works', async () => {
    const filePath = __dirname + '/data/data.qif';
    const { errors, transactions } = await parseFile(filePath);
    expect(errors.length).toBe(0);
    expect(transactions).toMatchSnapshot();
  });

  it('ofx import works', async () => {
    const filePath = __dirname + '/data/data.ofx';
    const { errors, transactions } = await parseFile(filePath);
    expect(errors.length).toBe(0);
    expect(transactions).toMatchSnapshot();
  }, 45000);

  it('qfx import works', async () => {
    const filePath = __dirname + '/data/data.qfx';
    const { errors, transactions } = await parseFile(filePath);
    expect(errors.length).toBe(0);
    expect(transactions).toMatchSnapshot();
  }, 45000);

  it('unsupported file type import fails', async () => {
    const filePath = __dirname + '/data/foo.txt';
    const { errors, transactions } = await parseFile(filePath);

    expect(errors.length).toBe(1);
    expect(errors[0].message).toBe('Invalid file type');
    expect(transactions.length).toBe(0);
  }, 45000);
});

describe('extractFileExtension', () => {
  it('matches extensions correctly (case-insensitive, etc)', () => {
    expect(extractFileExtension(__dirname + '/data/best.data-ever$.QFX')).toBe('.qfx');
    expect(extractFileExtension(__dirname + '/data/big.data.QiF')).toBe('.qif');

    // Unsupported filetype
    expect(extractFileExtension(__dirname + '/foo')).toBe(null);
  });

  it('returns null for invalid inputs', () => {
    expect(extractFileExtension(__dirname + '/foo')).toBe(null);
    expect(extractFileExtension('100')).toBe(null);
    expect(extractFileExtension(null)).toBe(null);
    expect(extractFileExtension(undefined)).toBe(null);
  });
});
