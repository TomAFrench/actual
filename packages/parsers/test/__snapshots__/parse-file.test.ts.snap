// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseFile csv import works 1`] = `
Array [
  Object {
    "amount": "-9.07",
    "date": "09/05/2022",
    "imported_payee": "THING",
    "notes": null,
    "payee_name": "THING",
  },
  Object {
    "amount": "-1000.00",
    "date": "09/05/2022",
    "imported_payee": "THING2",
    "notes": null,
    "payee_name": "THING2",
  },
  Object {
    "amount": "3012.12",
    "date": "06/05/2022",
    "imported_payee": "THING3",
    "notes": null,
    "payee_name": "THING3",
  },
]
`;

exports[`parseFile ofx import works 1`] = `
Array [
  Object {
    "amount": "-30.00",
    "date": "2019-01-23",
    "imported_id": "00092990122-30.00019012312798.01",
    "imported_payee": "PATIENT FIRST TOKEN 01/22 PURCHA",
    "notes": null,
    "payee_name": "PATIENT FIRST TOKEN 01/22 PURCHA",
  },
  Object {
    "amount": "-3.77",
    "date": "2019-01-23",
    "imported_id": "00092990121-3.77019012312828.01",
    "imported_payee": "STARBUCKS STORE 07604 01/21 PURC",
    "notes": null,
    "payee_name": "STARBUCKS STORE 07604 01/21 PURC",
  },
  Object {
    "amount": "-9.62",
    "date": "2019-01-23",
    "imported_id": "00092990121-9.62019012312831.78",
    "imported_payee": "STARBUCKS STORE 07604 01/21 PURC",
    "notes": null,
    "payee_name": "STARBUCKS STORE 07604 01/21 PURC",
  },
  Object {
    "amount": "-115.99",
    "date": "2019-01-22",
    "imported_id": "00090231800-115.99019012212841.40",
    "imported_payee": "VERIZON DES:PAYMENTREC ID:XXXXX3",
    "notes": null,
    "payee_name": "VERIZON DES:PAYMENTREC ID:XXXXX3",
  },
  Object {
    "amount": "-10.34",
    "date": "2019-01-22",
    "imported_id": "00092990120-10.34019012212957.39",
    "imported_payee": "URBAN FARMHOUSE NO 2 01/19 PURCH",
    "notes": null,
    "payee_name": "URBAN FARMHOUSE NO 2 01/19 PURCH",
  },
  Object {
    "amount": "-22.53",
    "date": "2019-01-22",
    "imported_id": "00092990120-22.53019012212967.73",
    "imported_payee": "URBAN FARMHOUSE NO 2 01/19 PURCH",
    "notes": null,
    "payee_name": "URBAN FARMHOUSE NO 2 01/19 PURCH",
  },
  Object {
    "amount": "-108.71",
    "date": "2019-01-22",
    "imported_id": "00092990119-108.71019012212990.26",
    "imported_payee": "TMOBILE*AUTO PAY 01/19 PURCHASE",
    "notes": null,
    "payee_name": "TMOBILE*AUTO PAY 01/19 PURCHASE",
  },
  Object {
    "amount": "-25.00",
    "date": "2019-01-22",
    "imported_id": "00092990118-25.00019012213098.97",
    "imported_payee": "COUNTY WASTE 01/18 PURCHASE 804-",
    "notes": null,
    "payee_name": "COUNTY WASTE 01/18 PURCHASE 804-",
  },
  Object {
    "amount": "-32.38",
    "date": "2019-01-22",
    "imported_id": "00092990118-32.38019012213123.97",
    "imported_payee": "REGENCY MART CITGO 01/18 PURCHAS",
    "notes": null,
    "payee_name": "REGENCY MART CITGO 01/18 PURCHAS",
  },
  Object {
    "amount": "-6.07",
    "date": "2019-01-22",
    "imported_id": "00092990117-6.07019012213156.35",
    "imported_payee": "CHICK-FIL-A #01342 01/17 PURCHAS",
    "notes": null,
    "payee_name": "CHICK-FIL-A #01342 01/17 PURCHAS",
  },
]
`;

exports[`parseFile qfx import works 1`] = `
Array [
  Object {
    "amount": "-30.00",
    "date": "2019-01-23",
    "imported_id": "00092990122-30.00019012312798.01",
    "imported_payee": "PATIENT FIRST TOKEN 01/22 PURCHA",
    "notes": null,
    "payee_name": "PATIENT FIRST TOKEN 01/22 PURCHA",
  },
  Object {
    "amount": "-3.77",
    "date": "2019-01-23",
    "imported_id": "00092990121-3.77019012312828.01",
    "imported_payee": "STARBUCKS STORE 07604 01/21 PURC",
    "notes": null,
    "payee_name": "STARBUCKS STORE 07604 01/21 PURC",
  },
  Object {
    "amount": "-9.62",
    "date": "2019-01-23",
    "imported_id": "00092990121-9.62019012312831.78",
    "imported_payee": "STARBUCKS STORE 07604 01/21 PURC",
    "notes": null,
    "payee_name": "STARBUCKS STORE 07604 01/21 PURC",
  },
  Object {
    "amount": "-115.99",
    "date": "2019-01-22",
    "imported_id": "00090231800-115.99019012212841.40",
    "imported_payee": "VERIZON DES:PAYMENTREC ID:XXXXX3",
    "notes": null,
    "payee_name": "VERIZON DES:PAYMENTREC ID:XXXXX3",
  },
  Object {
    "amount": "-10.34",
    "date": "2019-01-22",
    "imported_id": "00092990120-10.34019012212957.39",
    "imported_payee": "URBAN FARMHOUSE NO 2 01/19 PURCH",
    "notes": null,
    "payee_name": "URBAN FARMHOUSE NO 2 01/19 PURCH",
  },
  Object {
    "amount": "-22.53",
    "date": "2019-01-22",
    "imported_id": "00092990120-22.53019012212967.73",
    "imported_payee": "URBAN FARMHOUSE NO 2 01/19 PURCH",
    "notes": null,
    "payee_name": "URBAN FARMHOUSE NO 2 01/19 PURCH",
  },
  Object {
    "amount": "-108.71",
    "date": "2019-01-22",
    "imported_id": "00092990119-108.71019012212990.26",
    "imported_payee": "TMOBILE*AUTO PAY 01/19 PURCHASE",
    "notes": null,
    "payee_name": "TMOBILE*AUTO PAY 01/19 PURCHASE",
  },
  Object {
    "amount": "-25.00",
    "date": "2019-01-22",
    "imported_id": "00092990118-25.00019012213098.97",
    "imported_payee": "COUNTY WASTE 01/18 PURCHASE 804-",
    "notes": null,
    "payee_name": "COUNTY WASTE 01/18 PURCHASE 804-",
  },
  Object {
    "amount": "-32.38",
    "date": "2019-01-22",
    "imported_id": "00092990118-32.38019012213123.97",
    "imported_payee": "REGENCY MART CITGO 01/18 PURCHAS",
    "notes": null,
    "payee_name": "REGENCY MART CITGO 01/18 PURCHAS",
  },
  Object {
    "amount": "-6.07",
    "date": "2019-01-22",
    "imported_id": "00092990117-6.07019012213156.35",
    "imported_payee": "CHICK-FIL-A #01342 01/17 PURCHAS",
    "notes": null,
    "payee_name": "CHICK-FIL-A #01342 01/17 PURCHAS",
  },
]
`;

exports[`parseFile qif import works 1`] = `
Array [
  Object {
    "amount": -25,
    "date": "12/19/18",
    "imported_payee": "COUNTY WASTE 12/18 PURCHASE 804-8439288 VA",
    "notes": null,
    "payee_name": "COUNTY WASTE 12/18 PURCHASE 804-8439288 VA",
  },
  Object {
    "amount": -45,
    "date": "12/19/18",
    "imported_payee": "õSSELDORF KÖLN SALON - LIBBIE GR 12/18 PURCHASE RICHMOND VA",
    "notes": null,
    "payee_name": "õSSELDORF KÖLN SALON - LIBBIE GR 12/18 PURCHASE RICHMOND VA",
  },
  Object {
    "amount": -31.33,
    "date": "12/19/18",
    "imported_payee": "LOFT 0845 CHES 12/19 PURCHASE RICHMOND VA",
    "notes": null,
    "payee_name": "LOFT 0845 CHES 12/19 PURCHASE RICHMOND VA",
  },
  Object {
    "amount": -6.31,
    "date": "12/19/18",
    "imported_payee": "TARGET T- 1130 12/19 PURCHASE Richmond VA",
    "notes": null,
    "payee_name": "TARGET T- 1130 12/19 PURCHASE Richmond VA",
  },
  Object {
    "amount": -41.92,
    "date": "12/19/18",
    "imported_payee": "TARGET DEBIT CRD ACH TRAN CO ID:XXXXX15170 POS TARGET -1018 RICHMOND V",
    "notes": null,
    "payee_name": "TARGET DEBIT CRD ACH TRAN CO ID:XXXXX15170 POS TARGET -1018 RICHMOND V",
  },
  Object {
    "amount": -24.01,
    "date": "12/19/18",
    "imported_payee": "TARGET DEBIT CRD DES:ACH TRAN ID:XXXXX0370723991 INDN:S LONG 9321 CO I",
    "notes": null,
    "payee_name": "TARGET DEBIT CRD DES:ACH TRAN ID:XXXXX0370723991 INDN:S LONG 9321 CO I",
  },
  Object {
    "amount": -12,
    "date": "12/19/18",
    "imported_payee": "TARGET DEBIT CRD DES:ACH TRAN ID:XXXXX0370723991 INDN:S LONG 9703 CO I",
    "notes": null,
    "payee_name": "TARGET DEBIT CRD DES:ACH TRAN ID:XXXXX0370723991 INDN:S LONG 9703 CO I",
  },
  Object {
    "amount": 874.85,
    "date": "12/20/18",
    "imported_payee": "BKOFAMERICA MOBILE 12/20 3705340250 DEPOSIT *MOBILE VA",
    "notes": null,
    "payee_name": "BKOFAMERICA MOBILE 12/20 3705340250 DEPOSIT *MOBILE VA",
  },
  Object {
    "amount": -12.63,
    "date": "12/20/18",
    "imported_payee": "AMAZON.COM*M21KD0HA0 AM 12/19 PURCHASE AMZN.COM/BILL WA",
    "notes": null,
    "payee_name": "AMAZON.COM*M21KD0HA0 AM 12/19 PURCHASE AMZN.COM/BILL WA",
  },
  Object {
    "amount": -36,
    "date": "12/20/18",
    "imported_payee": "TARGET DEBIT CRD DES:ACH TRAN ID:XXXXX0370723991 INDN:S LONG 6486 CO I",
    "notes": null,
    "payee_name": "TARGET DEBIT CRD DES:ACH TRAN ID:XXXXX0370723991 INDN:S LONG 6486 CO I",
  },
  Object {
    "amount": -5.21,
    "date": "12/21/18",
    "imported_payee": "STARBUCKS STORE 07604 12/19 PURCHASE RICHMOND VA",
    "notes": null,
    "payee_name": "STARBUCKS STORE 07604 12/19 PURCHASE RICHMOND VA",
  },
  Object {
    "amount": -34.7,
    "date": "12/21/18",
    "imported_payee": "AMAZON.COM*M236431E2 AM 12/19 PURCHASE AMZN.COM/BILL WA",
    "notes": null,
    "payee_name": "AMAZON.COM*M236431E2 AM 12/19 PURCHASE AMZN.COM/BILL WA",
  },
  Object {
    "amount": -5.41,
    "date": "12/21/18",
    "imported_payee": "STARBUCKS STORE 07604 12/19 PURCHASE RICHMOND VA",
    "notes": null,
    "payee_name": "STARBUCKS STORE 07604 12/19 PURCHASE RICHMOND VA",
  },
  Object {
    "amount": -105.29,
    "date": "12/21/18",
    "imported_payee": "AMAZON.COM*M208P98U1 AM 12/19 PURCHASE AMZN.COM/BILL WA",
    "notes": null,
    "payee_name": "AMAZON.COM*M208P98U1 AM 12/19 PURCHASE AMZN.COM/BILL WA",
  },
  Object {
    "amount": -7.36,
    "date": "12/21/18",
    "imported_payee": "AMAZON.COM*M29QF7M71 AM 12/19 PURCHASE AMZN.COM/BILL WA",
    "notes": null,
    "payee_name": "AMAZON.COM*M29QF7M71 AM 12/19 PURCHASE AMZN.COM/BILL WA",
  },
  Object {
    "amount": -12.85,
    "date": "12/21/18",
    "imported_payee": "AMAZON.COM*M20VB1851 AM 12/20 PURCHASE AMZN.COM/BILL WA",
    "notes": null,
    "payee_name": "AMAZON.COM*M20VB1851 AM 12/20 PURCHASE AMZN.COM/BILL WA",
  },
  Object {
    "amount": -12.99,
    "date": "12/21/18",
    "imported_payee": "AMZN MKTP US*M27HH51C2 12/20 PURCHASE AMZN.COM/BILL WA",
    "notes": null,
    "payee_name": "AMZN MKTP US*M27HH51C2 12/20 PURCHASE AMZN.COM/BILL WA",
  },
  Object {
    "amount": -62.78,
    "date": "12/21/18",
    "imported_payee": "CVS/PHARM 0154 12/21 PURCHASE RICHMOND VA",
    "notes": null,
    "payee_name": "CVS/PHARM 0154 12/21 PURCHASE RICHMOND VA",
  },
]
`;
